packages/dataviews/src/dataforms-layouts/modal/index.tsx

```diff
@@ -0,0 +1,260 @@
+/**
+ * External dependencies
+ */
+import clsx from 'clsx';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	__experimentalVStack as VStack,
+	__experimentalHStack as HStack,
+	__experimentalSpacer as Spacer,
+	Button,
+	Modal,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { useState, useMemo, useContext } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import type { Form, FieldLayoutProps, SimpleFormField } from '../../types';
+import DataFormContext from '../../components/dataform-context';
+import { DataFormLayout } from '../data-form-layout';
+import { isCombinedField } from '../is-combined-field';
+
+function DataFormModalFooter( {
+	onCancel,
+	onApply,
+}: {
+	onClose: () => void;
+	onCancel: () => void;
+	onApply: () => void;
+} ) {
+	return (
+		<HStack className="dataforms-layouts-modal__modal-footer" spacing={ 3 }>
+			<Spacer />
+			<Button
+				variant="tertiary"
+				onClick={ onCancel }
+				__next40pxDefaultSize
+			>
+				{ __( 'Cancel' ) }
+			</Button>
+			<Button variant="primary" onClick={ onApply } __next40pxDefaultSize>
+				{ __( 'Apply' ) }
+			</Button>
+		</HStack>
+	);
+}
+
+function DataFormModal< Item >( {
+	isOpen,
+	onClose,
+	data,
+	form,
+	onChange,
+	fieldLabel,
+	onCancel,
+	onApply,
+}: {
+	isOpen: boolean;
+	onClose: () => void;
+	data: Item;
+	form: Form;
+	onChange: ( data: Partial< Item > ) => void;
+	onCancel: () => void;
+	onApply: () => void;
+	fieldLabel: string;
+} ) {
+	if ( ! isOpen ) {
+		return null;
+	}
+
+	return (
+		<Modal
+			className="dataforms-layouts-modal__modal"
+			onRequestClose={ onClose }
+			isFullScreen={ false }
+			title={ fieldLabel }
+			shouldCloseOnClickOutside={ false }
+		>
```

Comment:
Let’s use one of the preset `size` values. It might make sense to allow fields to pass a value along, but not essential for now.